### PR TITLE
Add infer for generic default type

### PR DESCRIPTION
**Pull request by edwin0cheng**
This PR add infer support for generic default type:

```
struct Gen<T=u32> {
    val: T
}
```

* add the (unresolved) defaults from the definition to GenericParams
* add a query generic_defaults that resolves those defaults to types and returns a Substs
* add the missing type in `substs_from_path_segment`
* add tests

based on the idea in this [comment](https://github.com/rust-analyzer/rust-analyzer/issues/1099#issuecomment-484206279)

### CODE REVIEW

File: crates/ra_hir/src/ty/lower.rs [hunk 1]
```diff
@@ -199,13 +201,28 @@ pub(super) fn substs_from_path_segment(
...
-    Substs(substs.into())
+    let mut substs = Substs(substs.into());
+
+    // handle defaults
+    if let Some(def_generic) = def_generic {
+        let default_substs = db.generic_defaults(def_generic);
+        assert_eq!(substs.len(), default_substs.len());
+
+        let mut i = 0;
+        substs.walk_mut(&mut |ty| {
```
Comment (flodiebold) [inline crates/ra_hir/src/ty/lower.rs]: walk_mut does a deep walk, you don't want that here. How about instead not turning the `Vec` into a `Substs` yet by leaving out line 209, then you can just do a normal for loop here?

Comment (edwin0cheng) [reply crates/ra_hir/src/ty/lower.rs]: Sure and Done :)

**Comment by flodiebold** [comment]
Thanks!

bors r+

**Pull request merged**